src/main/scala/fi/vm/sade/hakurekisteri/integration/VirkailijaRestClient.scala

```diff
@@ -263,12 +298,20 @@ object JsonExtractor extends HakurekisteriJsonSupport {
     }
     new FunctionHandler[T](f) {
       override def onCompleted(response: Res): T = {
-        if (codes.contains(response.getStatusCode))
+        if (codes.contains(response.getStatusCode) && !isRedirectToCasLogin(response))
           super.onCompleted(response)
-        else {
+        else if (isRedirectToCasLogin(response)){
+          throw new JSessionIdCookieException("JSESSIONID is expired")
+        } else {
           throw PreconditionFailedException(s"precondition failed for url: ${response.getUri}, status code: ${response.getStatusCode}, body: ${response.getResponseBody(Charset.forName("utf-8"))}", response.getStatusCode)
         }
       }
+
+      private def isRedirectToCasLogin(response: Res): Boolean = {
```

Comment:
Tässä varmaan riittäisi koko metodiksi jotain

```suggestion
      private def isRedirectToCasLogin(response: Res): Boolean = response.getUri.getPath.equals("/cas/login")
```